### PR TITLE
Change update configuration

### LINKED ISSUE #3086: Move all code relating to ministep/analysis configuration into submodule
We should move the configuration of ministeps and the analysis module into a seperate module, spesifically `LocalObsData` and `LocalMinistep`. Potentially also `local_config`.

### PR DESCRIPTION
**Pull request by oyvindeide**
**Issue**
Resolves #3086

**Approach**
Introduce Parameter and simplify the configuration. 

First commit:
Everything that seemed to be in use has been duplicated, and all the tests I could think of that configured the ministep passed (with the exception of one test in semeio where it checked the name of the obs data set, but that was only in test code, never done in production code).  There are some breaking changes in the configuration as somethings were seemingly only used in tests. It should be be trivial to achieve the same result as the removed functionality.

Second commit:
The old configuration is rewritten in pydantic, and breaks backwards compatibility. Removing everything from the old implementation.

## Pre review checklist

- [x] Added appropriate release note label
- [x] PR title captures the intent of the changes, and is fitting for release notes.
- [x] Commit history is consistent and clean, in line with the [contribution guidelines](https://github.com/equinor/ert/blob/main/CONTRIBUTING.md).

Adding labels helps the maintainers when writing release notes. This is the [list of release note labels](https://github.com/equinor/ert/labels?q=release-notes).

### CODE REVIEW

File: res/enkf/local_config.py [hunk 1]
```diff
@@ -13,120 +13,85 @@
...
-        # The grid can be None
-        return self.grid
-
-    def free(self):
-        self._free()
+@dataclass
+class UpdateConfiguration:
+    ministeps: List[LocalMinistep] = field(default_factory=list)
 
     def clear(self):
```
Comment (ManInFez) [inline res/enkf/local_config.py:28]: deprecate function

File: libres/lib/include/ert/enkf/enkf_analysis.hpp [hunk 2]
```diff
@@ -52,9 +51,8 @@ class UpdateSnapshot {
 UpdateSnapshot make_update_snapshot(const obs_data_type *obs_data,
                                     const meas_data_type *meas_data);
 
-void enkf_analysis_deactivate_outliers(obs_data_type *obs_data,
-                                       meas_data_type *meas_data,
-                                       double std_cutoff, double alpha,
-                                       bool verbose);
+void enkf_analysis_deactivate_outliers(
+    obs_data_type *obs_data, meas_data_type *meas_data, double std_cutoff,
+    double alpha, bool verbose, const std::vector<std::vector<int>> &obs_index);
```
Comment (ManInFez) [inline libres/lib/include/ert/enkf/enkf_analysis.hpp:56]: Pass in observation vector

File: libres/lib/include/ert/enkf/enkf_obs.hpp [hunk 3]
```diff
@@ -20,6 +20,8 @@
 #define ERT_ENKF_OBS_H
 #include <time.h>
 
+#include <ert/analysis/update.hpp>
```
Comment (ManInFez) [inline libres/lib/include/ert/enkf/enkf_obs.hpp:23]: please check include statement

File: libres/old_tests/enkf/test_obs_vector.cpp [hunk 4]
```diff
@@ -50,8 +41,8 @@ bool scale_std_summarysingleobservation_no_errors() {
     test_assert_double_equal(1.0, summary_obs_get_std_scaling(summary_obs));
 
     {
```
Comment (ManInFez) [inline libres/old_tests/enkf/test_obs_vector.cpp:43]: remove scope

File: tests/libres_tests/res/enkf/test_row_scaling_case.py [hunk 5]
```diff
@@ -568,26 +551,20 @@ def test_reuse_ALL_ACTIVE(self):
...
             update_node1 = EnkfNode(poro_config)
-            update_node2 = EnkfNode(poro_config)
             update_node3 = EnkfNode(poro_config)
             for iens in range(main.getEnsembleSize()):
                 node_id = NodeId(0, iens)
 
                 update_node1.load(update_fs1, node_id)
-                update_node2.load(update_fs2, node_id)
+
                 update_node3.load(update_fs3, node_id)
```
Comment (ManInFez) [inline tests/libres_tests/res/enkf/test_row_scaling_case.py:560]: Where is 2? rename 3 to 2

File: res/enkf/local_ministep.py [hunk 6]
```diff
@@ -1,84 +1,144 @@
...
+        current_version=__version__,
+    )
     def getName(self):
         """@rtype: str"""
         return self.name()
 
-    def free(self):
-        self._free()
+    def observation_config(self):
+        return [observation.to_tuple() for observation in self.observations.values()]
```
Comment (ManInFez) [inline res/enkf/local_ministep.py:126]: could this not be 
```suggestion
        return [(key, value) for observation in self.observations.items()]
```
or could the key in the dict and the name of observation differ?

Comment (ManInFez) [reply res/enkf/local_ministep.py:126]: This also removes the need of `to_tuple()` function

File: res/enkf/local_ministep.py [hunk 7]
```diff
@@ -1,84 +1,144 @@
...
+    def row_scaling(self, parameter_name) -> RowScaling:
+        if parameter_name in self.row_scaling_parameters:
+            return self.row_scaling_parameters[parameter_name].row_scaling
+        elif parameter_name in self.parameters:
+            parameter = self.parameters[parameter_name]
+            row_scaling = RowScaling()
+            self.row_scaling_parameters[parameter_name] = RowScalingParameter(
+                parameter.name, row_scaling, parameter.index_list
+            )
+            del self.parameters[parameter_name]
```
Comment (ManInFez) [inline res/enkf/local_ministep.py:97]: this delete is not self explained.. :shrug:

File: res/enkf/local_config.py [hunk 8]
```diff
@@ -13,120 +13,124 @@
...
+    @deprecated("Use add_ministep", removed_in="2.37", current_version=__version__)
+    def attachMinistep(self, mini_step):
+        if mini_step.name not in [ministep.name for ministep in self.ministeps]:
+            self.ministeps.append(mini_step)
 
-    def attachMinistep(self, update_step, mini_step):
-        assert isinstance(mini_step, LocalMinistep)
-        assert isinstance(update_step, LocalUpdateStep)
-        self._attach_ministep(update_step, mini_step)
+    def add_ministep(self, name):
```
Comment (ManInFez) [inline res/enkf/local_config.py:102]: Maybe ministeps should be created complete, i.e. take in a set of parameters and observations in the constructor. No partially valid objects

**Comment by dafeda** [comment]
> I think we should ta a look at the naming convention: One thing is analysis vs update. We are using the two terms interchangeably, I am not sure if it makes sense. We should aim to be a bit more consistent, at least in the code.
> 
> local-prefix. For me this is not very descriptive. I would try to find a something better, or remove prefix entirely
> 
> ministep, should be renamed to something descriptive.
> 
> I would call the whole hm-matching an analysis, and a smaller step an update-step and let the configuration naming follow. This might not be the best. @dafeda probably have some clever thoughts regarding this.

I've seen both `update` and `analyse` used in the literature, and I think I prefer `update` because I find `analyse` even more generic than `update`.

If we we want to use post-fix of `step`, I think I would have a `forecast-step`, `global-update-step` and `local-update-step`.

What do you think?

File: libres/lib/analysis/update.cpp [hunk 9]
```diff
@@ -495,6 +494,33 @@ static void run_analysis_update_with_rowscaling_pybind(
...
+                      &analysis::RowScalingParameter::get_index_list,
+                      &analysis::RowScalingParameter::set_index_list)
+        .def("__repr__", &::analysis::RowScalingParameter::to_string);
+
+    py::class_<analysis::Parameter, std::shared_ptr<analysis::Parameter>>(
+        m, "Parameter")
+        .def(py::init<std::string, std::vector<int>>(), py::arg("name"),
+             py::arg("index_list") = py::list())
+        .def_readwrite("name", &analysis::Parameter::name)
+        .def_readwrite("active_list", &analysis::Parameter::active_list)
```
Comment (ManInFez) [inline libres/lib/analysis/update.cpp:519]: make just readable

File: libres/lib/enkf/block_obs.cpp [hunk 10]
```diff
@@ -297,33 +282,19 @@ double block_obs_iget_data(const block_obs_type *block_obs, const void *state,
...
         int obs_size = block_obs_get_size(block_obs);
-        int active_size = __active_list->active_size(obs_size);
         meas_block_type *meas_block = meas_data_add_block(
             meas_data, block_obs->obs_key, node_id.report_step, obs_size);
-        int iobs;
 
-        active_mode_type active_mode = __active_list->getMode();
-        if (active_mode == ALL_ACTIVE) {
-            for (iobs = 0; iobs < obs_size; iobs++) {
+        {
```
Comment (ManInFez) [inline libres/lib/enkf/block_obs.cpp:292]: Please no more scopes than needed

File: libres/lib/enkf/enkf_analysis.cpp [hunk 11]
```diff
@@ -85,18 +85,34 @@ UpdateSnapshot make_update_snapshot(const obs_data_type *obs_data,
...
         obs_block_type *obs_block = obs_data_iget_block(obs_data, block_nr);
         meas_block_type *meas_block = meas_data_iget_block(meas_data, block_nr);
 
+        const std::vector<int> deactivate_index =
+            selected_obs.at(block_nr).second;
+        if (obs_block_get_key(obs_block) != selected_obs.at(block_nr).first)
+            throw std::invalid_argument(fmt::format(
+                "Expected obs_key: {}, got: {}", obs_block_get_key(obs_block),
+                selected_obs.at(block_nr).first));
         {
```
Comment (ManInFez) [inline libres/lib/enkf/enkf_analysis.cpp:103]: remove scope

File: libres/lib/enkf/enkf_obs.cpp [hunk 12]
```diff
@@ -296,33 +292,26 @@ int enkf_obs_get_size(const enkf_obs_type *obs) {
...
     while (true) {
         step = obs_vector_get_next_active_step(obs_vector, step);
         if (step < 0)
             break;
 
         const summary_obs_type *summary_obs =
             (const summary_obs_type *)obs_vector_iget_node(obs_vector, step);
-        double_vector_iset(obs_std, active_count,
+        observations.push_back(
+            std::make_pair(summary_obs_get_value(summary_obs),
```
Comment (ManInFez) [inline libres/lib/enkf/enkf_obs.cpp:312]:
```suggestion
        observations.push_back(
            {summary_obs_get_value(summary_obs), summary_obs_get_std(summary_obs) *
                               summary_obs_get_std_scaling(summary_obs))});
```

File: res/analysis/configuration.py [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+    ministeps: conlist(UpdateStep, min_items=1)
+
+    def __iter__(self):
+        yield from self.ministeps
+
+    def __len__(self):
+        return len(self.ministeps)
+
+    def __getitem__(self, item):
+        return self.ministeps[item]
```
Comment (ManInFez) [inline res/analysis/configuration.py:93]: Maybe not needed

File: res/enkf/enkf_main.py [hunk 14]
```diff
@@ -75,6 +76,8 @@ def __init__(self, config, strict=True):
         self._init_from_real_enkf_main(real_enkf_main)
         self._monkey_patch_methods(real_enkf_main)
 
+        self._local_config = None
```
Comment (ManInFez) [inline res/enkf/enkf_main.py:79]: maybe _update_conifg(uration)

File: res/analysis/configuration.py (hunk 13 above)
Comment (oyvindeide) [reply res/analysis/configuration.py:93]: Must have `len`, but have removed `getitem`

**Review by ManInFez — APPROVED** [review]
:bacon:

**Pull request merged**